### ISSUE TITLE
Can't add repos (https://github.com/mosaic5g/agent-r)

### ISSUE THREAD

**Issue by mosaic-5g** [canonical-websites/build.snapcraft.io#925]
The add repos is not finishing, and I am getting the following error from the js console:

```
api/launchpad/snaps/list?owner=mosaic5g Failed to load resource: the server responded with a status of 404 (Not Found)
api/launchpad/snaps/list?owner=mosaic5g Failed to load resource: the server responded with a status of 404 (Not Found)
bundle.6ef33c6d3bdf17491fd1.js:15 POST https://build.snapcraft.io/api/launchpad/snaps 400 (Bad Request)
(anonymous) @ bundle.6ef33c6d3bdf17491fd1.js:15
(anonymous) @ bundle.6ef33c6d3bdf17491fd1.js:71
(anonymous) @ bundle.6ef33c6d3bdf17491fd1.js:72
(anonymous) @ bundle.6ef33c6d3bdf17491fd1.js:55
dispatch @ bundle.6ef33c6d3bdf17491fd1.js:72
(anonymous) @ bundle.6ef33c6d3bdf17491fd1.js:24
o @ bundle.6ef33c6d3bdf17491fd1.js:72
(anonymous) @ bundle.6ef33c6d3bdf17491fd1.js:72
e.(anonymous function) @ bundle.6ef33c6d3bdf17491fd1.js:72
n @ bundle.6ef33c6d3bdf17491fd1.js:24
(anonymous) @ bundle.6ef33c6d3bdf17491fd1.js:24
Promise resolved (async)
n @ bundle.6ef33c6d3bdf17491fd1.js:24
(anonymous) @ bundle.6ef33c6d3bdf17491fd1.js:24
(anonymous) @ bundle.6ef33c6d3bdf17491fd1.js:24
(anonymous) @ bundle.6ef33c6d3bdf17491fd1.js:24
(anonymous) @ bundle.6ef33c6d3bdf17491fd1.js:24
o @ bundle.6ef33c6d3bdf17491fd1.js:72
(anonymous) @ bundle.6ef33c6d3bdf17491fd1.js:72
e.(anonymous function) @ bundle.6ef33c6d3bdf17491fd1.js:72
n @ bundle.6ef33c6d3bdf17491fd1.js:24
(anonymous) @ bundle.6ef33c6d3bdf17491fd1.js:24
(anonymous) @ bundle.6ef33c6d3bdf17491fd1.js:24
(anonymous) @ bundle.6ef33c6d3bdf17491fd1.js:24
(anonymous) @ bundle.6ef33c6d3bdf17491fd1.js:72
(anonymous) @ bundle.6ef33c6d3bdf17491fd1.js:55
value @ bundle.6ef33c6d3bdf17491fd1.js:52
n @ bundle.6ef33c6d3bdf17491fd1.js:25
a @ bundle.6ef33c6d3bdf17491fd1.js:25
s @ bundle.6ef33c6d3bdf17491fd1.js:25
f @ bundle.6ef33c6d3bdf17491fd1.js:22
m @ bundle.6ef33c6d3bdf17491fd1.js:22
r @ bundle.6ef33c6d3bdf17491fd1.js:47
processEventQueue @ bundle.6ef33c6d3bdf17491fd1.js:22
n @ bundle.6ef33c6d3bdf17491fd1.js:70
handleTopLevel @ bundle.6ef33c6d3bdf17491fd1.js:70
i @ bundle.6ef33c6d3bdf17491fd1.js:70
perform @ bundle.6ef33c6d3bdf17491fd1.js:23
batchedUpdates @ bundle.6ef33c6d3bdf17491fd1.js:70
i @ bundle.6ef33c6d3bdf17491fd1.js:9
dispatchEvent @ bundle.6ef33c6d3bdf17491fd1.js:70
bundle.6ef33c6d3bdf17491fd1.js:24 Uncaught (in promise) TypeError: Cannot read property 'id' of undefined
    at https://build.snapcraft.io/static/bundle.6ef33c6d3bdf17491fd1.js:24:3375
    at o (https://build.snapcraft.io/static/bundle.6ef33c6d3bdf17491fd1.js:72:24544)
    at Generator._invoke (https://build.snapcraft.io/static/bundle.6ef33c6d3bdf17491fd1.js:72:25946)
    at Generator.e.(anonymous function) [as throw] (https://build.snapcraft.io/static/bundle.6ef33c6d3bdf17491fd1.js:72:24723)
    at n (https://build.snapcraft.io/static/bundle.6ef33c6d3bdf17491fd1.js:24:2509)
    at https://build.snapcraft.io/static/bundle.6ef33c6d3bdf17491fd1.js:24:2634
    at <anonymous>
(anonymous) @ bundle.6ef33c6d3bdf17491fd1.js:24
o @ bundle.6ef33c6d3bdf17491fd1.js:72
(anonymous) @ bundle.6ef33c6d3bdf17491fd1.js:72
e.(anonymous function) @ bundle.6ef33c6d3bdf17491fd1.js:72
n @ bundle.6ef33c6d3bdf17491fd1.js:24
(anonymous) @ bundle.6ef33c6d3bdf17491fd1.js:24
Promise rejected (async)
n @ bundle.6ef33c6d3bdf17491fd1.js:24
(anonymous) @ bundle.6ef33c6d3bdf17491fd1.js:24
(anonymous) @ bundle.6ef33c6d3bdf17491fd1.js:24
(anonymous) @ bundle.6ef33c6d3bdf17491fd1.js:24
(anonymous) @ bundle.6ef33c6d3bdf17491fd1.js:72
(anonymous) @ bundle.6ef33c6d3bdf17491fd1.js:55
value @ bundle.6ef33c6d3bdf17491fd1.js:52
n @ bundle.6ef33c6d3bdf17491fd1.js:25
a @ bundle.6ef33c6d3bdf17491fd1.js:25
s @ bundle.6ef33c6d3bdf17491fd1.js:25
f @ bundle.6ef33c6d3bdf17491fd1.js:22
m @ bundle.6ef33c6d3bdf17491fd1.js:22
r @ bundle.6ef33c6d3bdf17491fd1.js:47
processEventQueue @ bundle.6ef33c6d3bdf17491fd1.js:22
n @ bundle.6ef33c6d3bdf17491fd1.js:70
handleTopLevel @ bundle.6ef33c6d3bdf17491fd1.js:70
i @ bundle.6ef33c6d3bdf17491fd1.js:70
perform @ bundle.6ef33c6d3bdf17491fd1.js:23
batchedUpdates @ bundle.6ef33c6d3bdf17491fd1.js:70
i @ bundle.6ef33c6d3bdf17491fd1.js:9
dispatchEvent @ bundle.6ef33c6d3bdf17491fd1.js:70
```

**Comment by bartaz** [comment]
@mosaic5g Thanks for the report. 

Your stack trace mentions `POST https://build.snapcraft.io/api/launchpad/snaps 400 (Bad Request)`.
Could you help us identify the issue by sending us the response body of this request.
This is an API request (you should see it in browser developer tools Network panel) and it should respond with JSON that would hopefully have some details about the error happening.

Thanks!

**Comment by kz6fittycent** [comment]
Can confirm same issue. 



```
2api/launchpad/snaps/list?owner=kz6fittycent Failed to load resource: the server responded with a status of 404 (Not Found)
api/launchpad/snaps Failed to load resource: the server responded with a status of 400 (Bad Request)
bundle.9c7c5495b89e55eca027.js:25 Uncaught (in promise) TypeError: Cannot read property 'id' of undefined
    at https://build.snapcraft.io/static/bundle.9c7c5495b89e55eca027.js:25:3477
    at o (https://build.snapcraft.io/static/bundle.9c7c5495b89e55eca027.js:73:24544)
    at Generator._invoke (https://build.snapcraft.io/static/bundle.9c7c5495b89e55eca027.js:73:25946)
    at Generator.e.(anonymous function) [as throw] (https://build.snapcraft.io/static/bundle.9c7c5495b89e55eca027.js:73:24723)
    at n (https://build.snapcraft.io/static/bundle.9c7c5495b89e55eca027.js:25:2610)
    at https://build.snapcraft.io/static/bundle.9c7c5495b89e55eca027.js:25:2735
    at <anonymous>
api/launchpad/snaps Failed to load resource: the server responded with a status of 400 (Bad Request)
```


and...
`
{"status":"error","payload":{"code":"lp-error","message":"url: Required input is missing."}}`

**Comment by bartaz** [comment]
@kz6fittycent Your repos/snaps are still visible in Launchpad (where they are built): https://launchpad.net/~build.snapcraft.io/+snaps/ - that's why trying to add them again fails.

Why they don't appear in 'My repos' is a different question. We had a similar issue once that was caused by problems with cache on server side.
We just recently deployed new version of BSI, so servers should be restarted and in fresh state.

Could you sign in again to BSI and see if you see your repos on My repos page?
And if not what are the network responses (are there any errors there)?

Thanks.

**Comment by kz6fittycent** [comment]
@bartaz I've logged out, cleared cache (in my browser), logged back in and I still don't show any repos available. Adding them is a no-go either. It's something that occurred during the Ubuntu hackathon in NYC. I assumed that my issues might be associated with higher loads and just dismissed it. 

However, here we are a few days afterward and I still can't add my repos back.

As a way forward, is there a way to remove my repos from Launchpad as a way to start fresh?

**Comment by bartaz** [comment]
@kz6fittycent Hi, thanks for checking it again.

Could you please see what error you get when you sign in and see no repos on https://build.snapcraft.io/user/kz6fittycent ?
Specifically there should be an API call to `https://build.snapcraft.io/api/launchpad/snaps/list?owner=kz6fittycent`. When everything is fine it should return 200 and JSON with your repos.
It seems that in your case it doesn't.

I checked the logs and I found a time when you visited and attempted to add a repo (that failed).
Surprisingly just before that in logs it's visible that BSI is calling GH API to fetch snapcraft.yaml from your repos. So it seems like they are fetched correctly from LP, by they get 'lost' before they reach you. If you could look at the error you get it may help us find the cause.

In theory - yes, it is possible to remove your repos from Launchpad (and BSI db, etc) to give you a fresh start for adding them. I'm just afraid that if we don't find the cause of the issue it may come back later (to you or some other user).

**Comment by bartaz** [comment]
@kz6fittycent I just had an idea and checked the list of your repos on Launchpad.

I found that 2 of the repositories don't exist: 
https://github.com/kz6fittycent/telegram-snap
https://github.com/kz6fittycent/sayonara-snap

I suppose you added them at some time and then removed from GH (without removing them from BSI). Which is a valid use case and it shouldn't cause the problem.
But I suspect that it's the cause: before we show you your repos we are trying to fetch some of their data from GH and it fails for repos that don't exist and it seems that makes whole response ending up as an error.

I'll remove just these 2 repos (that don't exist in GH) from Launchpad and we can check later if that helps.

**Comment by bartaz** [comment]
@kz6fittycent I think I successfully removed these two removed repos from Launchpad.
Hopefully this will let you see your other repos correctly without adding them again.

**Comment by bartaz** [comment]
I verified it myself (added a repo and removed it from GH) and I can confirm that this causes this bug.

Reported it separately as #968

**Comment by kz6fittycent** [comment]
@bartaz Those repos were never actually added in GH. They were snaps I attempted to build locally and registered the name (as I was following the steps listed in the "build your first snap" guide). 

BTW, your idea worked! I can now see my snaps and things are back in swing again. 

Thanks for your help!

**Comment by bartaz** [comment]
Fixed by #973
